Python: Add reasoning support for OpenAI Responses Agents (GPT-5, o4-mini, o3)

ltwlf opened the pull request:
## Summary

Adds reasoning capabilities to Python OpenAI ResponsesAgent, bringing parity with the C# implementation. This enables fine-grained control over reasoning effort for O-series models (o1, o3-mini, o4-mini) and gpt-5 that support the reasoning parameter.

Fixes #12843

## Changes

### Core Implementation
- **`openai_responses_agent.py`**: Added reasoning configuration support
  - Constructor-level reasoning effort setting
  - Per-invocation reasoning effort override capability
  - Proper parameter validation and model compatibility checks
- **`responses_agent_thread_actions.py`**: Extended thread actions to support reasoning parameters
  - Reasoning effort propagation through thread operations
  - Metadata preservation for reasoning tokens and summaries

### Sample and Tests
- **`responses_agent_reasoning.py`**: Comprehensive demonstration sample
  - Constructor vs per-invocation reasoning configuration
  - Function calling integration with reasoning
  - Reasoning comparison scenarios (low/medium/high effort)
  - Error handling and troubleshooting guidance
- **`test_openai_responses_agent_reasoning.py`**: Full unit test coverage
  - Parameter validation tests
  - Integration scenarios with function calling
  - Edge cases and error conditions

## Features

✅ **Constructor-Level Reasoning**: Set default reasoning effort when creating agents  
✅ **Per-Invocation Override**: Override reasoning effort per request  
✅ **Priority Hierarchy**: per-invocation > constructor > model default  
✅ **Function Calling Compatible**: Works seamlessly with existing plugin system  
✅ **Azure OpenAI & OpenAI Support**: Compatible with both service providers  
✅ **Model Validation**: Automatic compatibility checks for O-series models  nad GÜT-5
✅ **Metadata Access**: Reasoning tokens and summaries available in response metadata  

## Usage Example

```python
# Constructor-level reasoning
# Constructor-level reasoning configuration
agent = AzureResponsesAgent(
    ai_model_id="gpt-5",
    client=client,
    reasoning={"effort": "low"}  # Default reasoning for all requests
)

# Per-invocation override
response = await agent.invoke(
    "Solve this complex problem step by step",
    reasoning={"effort": "high"}  
)


# Invoke with reasoning callback to capture intermediate thoughts
response = await agent.invoke(
    "Analyze this data step by step",
    reasoning={"effort": "high", "summary": "detailed"},
    on_intermediate_message=handle_reasoning_message
)

# Streaming with reasoning
async for response in agent.invoke_stream(
    "Explain quantum computing in detail",
    reasoning={"effort": "high", "summary": "detailed"},
    on_intermediate_message=handle_reasoning_message
):
    print(response.content, end="", flush=True)

moonbox3 commented:
<!-- Pytest Coverage Comment: python-test-coverage-report -->
<a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/README.md"><img alt="Python Test Coverage" src="https://img.shields.io/badge/Python Test Coverage-82%25-green.svg" /></a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>agents/open_ai</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py">openai_responses_agent.py</a></td><td>422</td><td>104</td><td>75%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L66">66</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L95">95</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L114-L117">114&ndash;117</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L125">125</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L129-L130">129&ndash;130</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L162-L166">162&ndash;166</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L170">170</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L175">175</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L180">180</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L185">185</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L191">191</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L196-L197">196&ndash;197</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L200">200</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L203">203</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L208-L213">208&ndash;213</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L218-L219">218&ndash;219</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L221-L222">221&ndash;222</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L228-L231">228&ndash;231</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L237">237</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L239-L240">239&ndash;240</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L244-L248">244&ndash;248</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L341">341</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L345">345</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L349">349</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L351">351</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L357">357</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L370">370</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L372">372</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L374">374</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L378">378</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L497-L498">497&ndash;498</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L505-L506">505&ndash;506</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L508-L509">508&ndash;509</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L511-L512">511&ndash;512</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L514-L519">514&ndash;519</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L521">521</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L569-L570">569&ndash;570</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L573">573</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L618">618</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L634">634</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L644">644</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L678">678</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L686">686</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L689">689</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L693">693</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L695">695</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L698">698</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L721">721</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L724-L728">724&ndash;728</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L738">738</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L778">778</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L803">803</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L814">814</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L905">905</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L1028">1028</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L1072">1072</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L1074">1074</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L1148">1148</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L1191-L1192">1191&ndash;1192</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L1194-L1198">1194&ndash;1198</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/openai_responses_agent.py#L1217">1217</a></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py">responses_agent_thread_actions.py</a></td><td>398</td><td>102</td><td>74%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L172">172</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L189">189</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L200">200</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L211-L212">211&ndash;212</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L382-L384">382&ndash;384</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L398">398</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L403">403</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L410">410</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L428-L429">428&ndash;429</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L433">433</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L477-L479">477&ndash;479</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L482">482</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L500-L501">500&ndash;501</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L510-L512">510&ndash;512</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L516">516</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L520-L521">520&ndash;521</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L525-L527">525&ndash;527</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L538-L540">538&ndash;540</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L562">562</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L564-L565">564&ndash;565</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L572-L573">572&ndash;573</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L575-L576">575&ndash;576</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L578">578</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L580-L581">580&ndash;581</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L583-L584">583&ndash;584</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L682">682</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L687">687</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L695-L696">695&ndash;696</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L698">698</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L704">704</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L709">709</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L714-L715">714&ndash;715</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L722-L724">722&ndash;724</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L729-L730">729&ndash;730</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L732-L737">732&ndash;737</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L739">739</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L743-L744">743&ndash;744</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L746">746</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L750">750</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L754">754</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L759-L760">759&ndash;760</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L765">765</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L768-L769">768&ndash;769</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L840-L843">840&ndash;843</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L884-L885">884&ndash;885</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L889-L891">889&ndash;891</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L893">893</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L918-L919">918&ndash;919</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L921-L925">921&ndash;925</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L927">927</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L937-L938">937&ndash;938</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L946">946</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L950">950</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L957">957</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L963">963</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L1070">1070</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L1093">1093</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/41530913d0260eac544c3301906b2d1047332749/agents/open_ai/responses_agent_thread_actions.py#L1143">1143</a></td></tr><tr><td><b>TOTAL</b></td><td><b>26835</b></td><td><b>4585</b></td><td><b>82%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 3699 | 22 :zzz: | 0 :x: | 0 :fire: | 1m 55s :stopwatch: |

ltwlf commented:
GPT-5 reasoning implementation tested and working flawlessly!

Key findings:
- GPT-5 supports reasoning with reasoning_effort="low" for structured responses
- Works great without reasoning too
- Function calling performs excellently with calculator demos

Model behavior:
- O-series (o1, o3-mini, o4-mini): Auto-reasoning when unspecified
- GPT-5: Reasoning support with new minimal effort option
- GPT-4.1: Fails with reasoning params (expected)

Implementation notes:
- Priority hierarchy handles GPT-5 seamlessly
- OpenAIResponsesAgent(ai_model_id="gpt-5") works out of the box
- Explicit reasoning leverages GPT-5's enhanced capabilities

Result: Reasoning feature is ready for OpenAI's latest models and gracefully handles both current O-series and future reasoning-capable models.

ltwlf commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -98,7 +100,7 @@ async def invoke(
         parallel_tool_calls: bool | None = None,
         polling_options: "RunPollingOptions | None" = None,
-        reasoning: Literal["low", "medium", "high"] | None = None,
+        reasoning: Literal["minimal", "low", "medium", "high"] | None = None,
```
The OpenAI API specification supports 'minimal' as a valid reasoning effort level. Our implementation correctly includes it.

ltwlf commented:
I've also added GPT-5 verbosity param

moonbox3 commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -98,12 +100,13 @@ async def invoke(
         parallel_tool_calls: bool | None = None,
         polling_options: "RunPollingOptions | None" = None,
-        reasoning: Literal["low", "medium", "high"] | None = None,
+        reasoning: Literal["minimal", "low", "medium", "high"] | None = None,
```
I think this `reasoning` literal string should be replaced with OpenAI's `Reasoning` typed dict (from openai.types.shared_params.reasoning import Reasonin).

ltwlf replied:
@moonbox3 Ok, but "minimal" was introduced 5 days ago with GPT-5. Can you please bump openai to > 1.99.8 to have minimal effort.

ltwlf replied:
done

moonbox3 commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -311,11 +321,16 @@ async def invoke_stream(
             An async iterable of tuple of the visibility of the message and the chat message content.
         """
+        # Validate reasoning effort parameter
+        cls._validate_reasoning_effort_parameter(reasoning)
```
I don't think we need to be so defensive with this validation check.

ltwlf replied:
removed

moonbox3 commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -653,13 +669,25 @@ def _prepare_chat_history_for_request(
+                and message.inner_content.output
+            ):
+                # Pass through raw output items directly to preserve API compliance
+                response_inputs.extend(message.inner_content.output)
```
Not sure I follow why this inner_content / inner_content.output attribute check was introduced.

ltwlf replied:
This was a pass through to fix function calling, but u resolved this better with PR #12907.

moonbox3 commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -705,22 +735,16 @@ def _prepare_chat_history_for_request(
                             "call_id": content.call_id,
                         }
                         response_inputs.append(rfrc_dict)
+                    case FunctionCallContent():
```
This isn't the case when `store_enabled` is False. We need to build up the FCC based on the chat history. Please revert this change, I'm handling it in a different PR.

moonbox3 commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -793,6 +817,19 @@ def _get_tool_calls_from_output(
         return function_calls
 
+    @classmethod
+    def _get_reasoning_items_from_output(
```
Response reasoning output consists of interim messages from the model, before the answer is delivered. We'll want to make sure these items are delivered to the caller via the `on_intermediate_message` callback. Our goal is to keep the `invoke` and `invoke_stream` methods delivering "final responses" with interim messages handled via the callback.

ltwlf commented:
@moonbox3 thanks for the thorough review! I'll work through the feedback and should have everything addressed by Wednesday

ltwlf commented on python/semantic_kernel/agents/open_ai/openai_responses_agent.py:
```diff
@@ -374,10 +379,60 @@ def __init__(
+        return self._default_reasoning_effort
+
+    @property
+    def verbosity_level(self) -> str | None:
```
Indeed, not necessary - removed

ltwlf commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -982,27 +1021,154 @@ def _merge_options(
+        return None
+
+    @classmethod
+    def _is_o_series_model(cls: type[_T], model_name: str) -> bool:
```
@eavanvalkenburg It doesn't conflict with GPT-5 - it just establishes a default configuration for the o-series models. I actually use this branch with GPT-5 both with and without reasoning enabled.
@moonbox3 I wasn't entirely sure if this approach still makes sense going forward. I think the o-series models will gradually be phased out as GPT-5 can be used with and without reasoning. I'll remove it.

ltwlf commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -982,27 +1021,154 @@ def _merge_options(
+        verbosity = merged.get("verbosity", None)
+
+        # Apply reasoning effort priority hierarchy: per-invocation > constructor > model default
+        effective_reasoning = cls._resolve_reasoning_effort(
```
@moonbox3  I use this in multi-agent scenarios with group chat to configure individual agent defaults. Setting defaults per model would affect all agents, and per-invocation configuration isn't ideal for group chat managers.
Or what would be a better approach?

eavanvalkenburg commented on python/semantic_kernel/contents/reasoning_content.py:
```diff
@@ -0,0 +1,89 @@
+_T = TypeVar("_T", bound="ReasoningContent")
+
+
+class ReasoningContent(KernelContent):
```
this is fully designed for the OpenAI api, which is not what we want to do...

ltwlf commented:
@angangwa yes, I'm just working on this!

ltwlf commented on python/semantic_kernel/agents/open_ai/responses_agent_thread_actions.py:
```diff
@@ -982,27 +1021,154 @@ def _merge_options(
+        return model_lower.startswith("o1") or model_lower.startswith("o3") or model_lower.startswith("o4")
+
+    @classmethod
+    def _get_default_reasoning_for_model(cls: type[_T], model_name: str) -> str:
```
done

ltwlf commented on python/semantic_kernel/agents/open_ai/openai_responses_agent.py:
```diff
@@ -374,10 +379,60 @@ def __init__(
+        Raises:
+            AgentInitializationException: If the verbosity level is invalid.
+        """
+        if verbosity is not None and verbosity not in ["low", "medium", "high"]:
```
removed. 
I want to reintroduce verbosity param after openai sdk bump to latest

ltwlf commented on python/semantic_kernel/agents/open_ai/openai_responses_agent.py:
```diff
@@ -374,10 +379,60 @@ def __init__(
         super().__init__(**args)
 
+        # Validate and store agent-level reasoning effort
+        self._validate_reasoning_effort(reasoning_effort)
```
done

ltwlf commented on python/semantic_kernel/contents/reasoning_content.py:
```diff
@@ -0,0 +1,89 @@
+_T = TypeVar("_T", bound="ReasoningContent")
+
+
+class ReasoningContent(KernelContent):
```
You are absolutely right.
I've refactored this and also added a steaming content type.
Done

ltwlf commented on python/tests/unit/agents/open_ai/test_openai_responses_agent_reasoning.py:
```diff
@@ -0,0 +1,347 @@
+from semantic_kernel.exceptions.agent_exceptions import AgentInitializationException
+
+
+class TestOpenAIResponsesAgentReasoning:
```
Done

ltwlf commented:
Hi @moonbox3 @eavanvalkenburg,

I've refactored the code and implemented the suggested changes with one exception: I've kept the constructor-level reasoning parameter. I believe this is useful for multi-agent collaboration scenarios where an orchestrator automatically invokes agents. 
If you think otherwise I can change it.
It would be interesting to have a smarter orchestrator that can decide what reasoning effort is needed based on context, but that's beyond the scope of this PR.

I've also removed the "minimal" effort option since I'm now using OpenAI types directly. The minimal effort will automatically return when we update to the latest OpenAI SDK.

I've tested this implementation with O-series models and GPT-5. After this PR is merged, I plan to bump the OpenAI version to the latest and then submit another PR adding GPT-5 verbosity support.

Could you please review the changes when you have a chance?

dmytrostruk commented on python/semantic_kernel/contents/reasoning_content.py:
```diff
@@ -0,0 +1,48 @@
+class ReasoningContent(KernelContent):
+    """Represents reasoning content.
+
+    Mirrors the .NET ReasoningContent which only exposes text. Any provider-specific
```
I'm not sure if `.NET` mentioning here is useful for end-users. Let's update this info in a way that explains what the purpose of this type is, so end-users can understand it better. In similar way how it's done for `StreamingReasoningContent` class.

moonbox3 commented on python/samples/concepts/agents/openai_responses/responses_agent_reasoning.py:
```diff
@@ -0,0 +1,241 @@
+    return None, None
+
+
+async def create_reasoning_agent():
```
There's a lot going on in this sample. I think it's totally fine to scope the samples to as small as possible. Meaning, we have one sample for `reasoning_with_summaries` and one sample for `reasoning`. Additionally, we try to break up the `invoke` case versus the `streaming` case.

moonbox3 commented on python/samples/concepts/agents/openai_responses/responses_agent_reasoning.py:
```diff
@@ -0,0 +1,241 @@
+        print("The reasoning summary feature is only available to verified OpenAI organizations.")
+
+    print("\n" + "=" * 60)
+    print("Demo complete! Key differences:")
```
This is unnecessary GitHub Copilot logging, can we remove, please?

moonbox3 commented on python/semantic_kernel/agents/open_ai/openai_responses_agent.py:
```diff
@@ -190,7 +191,6 @@ def id(self) -> str | None:
         """Get the thread ID."""
         return self.response_id
 
-    @override
```
`_create` is an abstract method in the base class - we should leave the `@override` decorator.

moonbox3 commented on python/semantic_kernel/agents/open_ai/openai_responses_agent.py:
```diff
@@ -826,12 +838,12 @@ async def get_response(
         temperature: float | None = None,
         top_p: float | None = None,
         truncation: str | None = None,
+        reasoning: Reasoning | dict[str, Any] | None = None,
```
Previous parameters were alphabetized:

...
`polling_options`
`reasoning`
`text`
...

moonbox3 commented:
@ltwlf please also run `mypy` locally, there are some failures showing up in the code quality checks. 

https://github.com/microsoft/semantic-kernel/actions/runs/16966028610/job/48510521071?pr=12881

moonbox3 commented:
> > @ltwlf It looks like there are some code quality check failures in CI.
> 
> @dmytrostruk not sure what was wrong. Pre-commit checks were successful on my box: <img alt="image" width="560" height="179" src="https://private-user-images.githubusercontent.com/965766/480581120-0164a44d-1b0c-49df-b6d9-efa8980a9fb6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYzNjU3MTAsIm5iZiI6MTc1NjM2NTQxMCwicGF0aCI6Ii85NjU3NjYvNDgwNTgxMTIwLTAxNjRhNDRkLTFiMGMtNDlkZi1iNmQ5LWVmYTg5ODBhOWZiNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyOFQwNzE2NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMzUxNDM2MmNjNDBkNjIzZmMyOTg0YjU2YzVkNzBjNDc4MjMzMWVjNTNhZThhNmY0ZWMwZjFhODYzY2MxYmI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.8uH5Q7RcMdcLoqchNz7WoG5hCOSgRyal8pcEmI9TwLg">
> 
> I've resolved the review comments. Hope it will be fine now 😅

Mypy isn’t run as part of pre-commit. It’s run separately in CI/CD. That’s why one should manually run the mypy task locally.

ltwlf commented:
> @ltwlf do you want me to fix these mypy issues?
Yes, please—if you don’t mind.
Not sure why mypy works locally for me but not for others. Maybe I should reset my project setup
